src/FreeAge/common/type_info.cpp

```diff
@@ -0,0 +1,621 @@
+      // TODO: implement
+    }
+  }
+
+  // TODO: add the ApplyToTechnology
+
+private:
+
+  void ApplyToObjectType(ObjectTypeStats* stats, const ObjectTypeStats* statsBase) const {
+    // TODO: Change the given stats
+
+    switch(type) {
+
+      CalculateCaseI(Type::MaxHp, maxHp)
+      CalculateCaseF(Type::MaxRange, maxRange)
+      CalculateCaseF(Type::FireRate, fireRate)
+
+      case Type::Cost: {
+        for (int i=0; i<4; i++) {
+          if (extra == Unset || extra == i) {
+            // TODO: extend the ResourceAmounts interface
+            // stats->cost.Set(i, CalculateInt(stats->cost.Get(i), statsBase->cost.Get(i));
+          }
+        }
+        break;
+      }
+      // TODO: implement all
+    }
+  }
+
+  int CalculateInt(int current, int base) const {
+    switch(operation) {
+    case Operation::Set: return value;
+    case Operation::Add: return current + value;
+    case Operation::Mult: return current * 100.0f / value;
+    case Operation::MultAdd: return current + base * 100.0f / value;
+    }
+  }
+  float CalculateFloat(float current, float base) const {
+    switch(operation) {
+    case Operation::Set: return value;
+    case Operation::Add: return current + value;
+    case Operation::Mult: return current * 100.0f / value;
+    case Operation::MultAdd: return current + base * 100.0f / value;
+    }
+  }
+
+};
+
+
+struct ObjectTypeFilter {
+
+  enum class Type {
+    AllUnits,
+    UnitByType,
+    UnitByArmor,
+    AllBuildings,
+    BuildingByType,
+    BuildingByArmor,
+    Technology,
+    Civilization
+  };
+
+  Type type;
+  int data; // UnitType, BuildingType or DamageType (use union ?)
+
+  bool Matches(/* TODO: implement base on the use case */) const;
+};
+
+ObjectTypeFilter FilterOfUnitsByType(UnitType type);
+ObjectTypeFilter FilterOfBuildingsByType(BuildingType type);
+ObjectTypeFilter FilterOfUnitsByArmor(DamageType type);
+ObjectTypeFilter FilterOfBuildingsByArmor(DamageType type);
+ObjectTypeFilter FilterOfTechnology(TechnologyType type);
+
+
+struct FilteredModification {
+  ObjectTypeFilter filter;
+  Modification modication;
+};
+
+enum class TechnologyType {
+  Loom,
+  // ...
+  DarkAge
+  // ...
+  // TODO: fill
+};
+
+struct TechnologyStats {
+
+  // TODO: implement a copy constructor to deep copy the modications vector,
+  //       should behave like the modications is a static array in order to
+  //       behave the same like the other Stats.
+
+  ResourceAmount cost;
+  float researchTime;
+
+  std::vector<FilteredModification> modications;
+
+  void addModification(FilteredModification x);
+
+};
+
+typedef TechnologyStats AgeStats;
+
+
+// ===================================================================================
+// === Usage examples
+// ===================================================================================
+
+// Technology
+void test1() {
+
+  TechnologyStats* loom = civilization.AddTechology(TechnologyType::Loom); // alocated inside a vector, return pointer
```

Comment:
Had this in mind... 

> vector can be pre-allocated with the final size

this is what I mean by `// alocated inside a vector`, don't know how to call this?

`AddTechology` make no sense either way, it would be something like that `TechnologyStats* loom = &civilization.techologies.at(TechologyType::Loom);` (or with a reference?)

Or even use a array?

Reply:
"allocated inside a vector" made me think "it returns the address to an element allocated with `push_back()`". Something like `civilization.techologies.at` makes it clear to me that it does not actually allocate something. I think that the choice of vector or array probably does not make a practical difference here.